src/Orleans/Providers/ILeaseProvider.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Orleans.LeaseProviders
+{
+    /// <summary>
+    /// Acquired lease
+    /// </summary>
+    public class AcquiredLease 
+    {
+        /// <summary>
+        /// The resource key which the lease is attached to 
+        /// </summary>
+        public string ResourceKey { get; }
+        /// <summary>
+        /// Duration of the aquired lease
+        /// </summary>
+        public TimeSpan Duration { get; }
+        /// <summary>
+        /// Lease token, whcih will be null if acquiring or renewing the lease failed
+        /// </summary>
+        public string Token { get; }
+        /// <summary>
+        /// Start time for this lease, which is when the lease is acquired or renewed
+        /// </summary>
+        public DateTime StartTimeUtc { get; }
+    }
+
+    /// <summary>
+    /// AcquireLeaseResult class, which demonstrates result of acquiring or renewing lease operation
+    /// </summary>
+    public class AcquireLeaseResult
+    {
+        /// <summary>
+        /// Acquired lease, which will be null if acquire or renew operation failed.
+        /// </summary>
+        AcquiredLease AcquiredLease { get; }
+        /// <summary>
+        /// Whether successfully acquire or renew the lease
+        /// </summary>
+        public bool Success { get; }
+        /// <summary>
+        /// If acquiring or renewing the lease failed, this is the exception which caused it. This field would be null if operation succeed.
+        /// </summary>
+        public Exception FailureException { get; }
```

Comment:
Do we want to define exceptions as part of the public api.  I ask because a timeout exception on a renew does not mean that the lease has been lost, so renew can be retried.  Similarly, a timeout or storage failure on an acquire would also indicate it can be retried.  Seems like exceptions should fall into LeaseNotAvailableException meaning the lease is owned by another (should not retry), NotLeaseOwnerException which would indicate we're trying to renew or release a lease we no longer own (should not be retried), and LeaseActionFailedException which is a catch all for all other lease errors like timeout, storage, and the like (can be retried).
Without well defined exceptions, I don't think there is enough information in the results for uses to correctly handle errors.